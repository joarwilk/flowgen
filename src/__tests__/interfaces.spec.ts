import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle single interface", () => {
  const ts = `
interface User {
  firstName: string
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
  const result2 = compiler.compileDefinitionString(ts, {
    interfaceRecords: true,
  });
  expect(beautify(result2)).toMatchSnapshot();
  expect(result2).toBeValidFlowTypeDeclarations();
});

it("should handle interface inheritance", () => {
  const ts = `
interface User {
  firstName: string
}
interface SpecialUser extends User {
  nice: number
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
  const result2 = compiler.compileDefinitionString(ts, {
    interfaceRecords: true,
  });
  expect(beautify(result2)).toMatchSnapshot();
  expect(result2).toBeValidFlowTypeDeclarations();

  const result3 = compiler.compileDefinitionString(ts, {
    interfaceRecords: true,
    inexact: false,
  });
  expect(beautify(result3)).toMatchSnapshot();
  expect(result3).toBeValidFlowTypeDeclarations();

  const result4 = compiler.compileDefinitionString(ts, {
    inexact: false,
  });
  expect(beautify(result4)).toMatchSnapshot();
  expect(result4).toBeValidFlowTypeDeclarations();
});

it("should handle interface merging", () => {
  const ts = `
interface User {
  firstName: string
}
interface User {
  lastName: string
}
interface User {
  username: string
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
  const result2 = compiler.compileDefinitionString(ts, {
    interfaceRecords: true,
  });
  expect(beautify(result2)).toMatchSnapshot();
  expect(result2).toBeValidFlowTypeDeclarations();
});

it("should handle all properties", () => {
  const ts = `
interface Props {
  "aria-label": string;
  "aria-labelledby"?: number;
  color: string;
  [key: string]: string;
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).not.toBeValidFlowTypeDeclarations(); // unsupported-syntax
});

it("should support readonly modifier", () => {
  const ts = `
interface Helper {
  readonly name: string;
  readonly callback(): void;
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should support call signature", () => {
  const ts = `
  interface ObjectSchema<T> {}
  interface ObjectSchemaDefinition<T> {}
  declare interface ObjectSchemaConstructor {
    <T extends object>(fields?: ObjectSchemaDefinition<T>): ObjectSchema<T>;
    new (): ObjectSchema<{}>;
  }
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should remove this in call signature", () => {
  const ts = `
interface Arc<This, Datum> {
  (this: This, d: Datum, ...args: any[]): string | null;
}
  
interface D<This, Datum> {
  new (this: This, d: Datum, ...args: any[]);
}
  
interface C<This, Datum> {
  (this: This, d: Datum, ...args: any[]);
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should remove generic defaults in call signature", () => {
  const ts = `
interface AbstractLevelDOWN<K, V> {}
interface AbstractLevelDOWNConstructor {
    <K = any, V = any>(location: string): AbstractLevelDOWN<K, V>;
}  
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should support omitting generic defaults in types, classes, interfaces", () => {
  const ts = `
interface Foo<T = symbol, U = number> {}
interface FooBar extends Foo {}
type Bar<T = number, U = string> = {}
class Baz<T = string, U = number> {}

declare var a: Foo
declare var b: Bar
declare var c: Baz

declare var d: Foo<any>
declare var e: Bar<any>
declare var f: Baz<any>
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should support optional methods", () => {
  const ts = `
interface Example<State> {
  required<R>(value: any, state: State): true;
  optional?<R>(value: any, state: State): false;
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle toString property name", () => {
  const ts = `
interface A {
  toString(): string;
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle untyped object binding pattern", () => {
  const ts = `
interface ObjectBinding {
  (): void;
  ({}): void;
  ({ a, b }): void;
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle untyped array binding pattern", () => {
  const ts = `
interface ArrayBinding {
  (): void;
  ([]): void;
  ([ a, b ]): void;
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle typed object binding pattern", () => {
  const ts = `
interface ObjectBinding {
  (): void;
  ({}: any): void;
  ({ a, b }: { a: string, b: number }): void;
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle typed array binding pattern", () => {
  const ts = `
interface ArrayBinding {
  (): void;
  ([]: []): void;
  ([ a, b ]: [string, number]): void;
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle mutli-extends pattern", () => {
  const ts = `
interface Shape {
  color: string;
}

interface PenStroke {
  penWidth: number;
}
interface Square extends Shape, PenStroke {
  sideLength: number;
}
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});
