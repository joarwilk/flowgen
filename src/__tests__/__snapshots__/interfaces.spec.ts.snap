// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle all properties 1`] = `
"declare interface Props {
  \\"aria-label\\": string;
  \\"aria-labelledby\\"?: number;
  color: string;
  [key: string]: string;
}
"
`;

exports[`should handle interface inheritance 1`] = `
"declare interface User {
  firstName: string;
}
declare type SpecialUser = { nice: number, ... } & User;
"
`;

exports[`should handle interface inheritance 2`] = `
"declare type User = { firstName: string, ... };
declare type SpecialUser = {
  ...$Exact<User>,
  nice: number,
  ...
};
"
`;

exports[`should handle interface inheritance 3`] = `
"declare type User = {| firstName: string |};
declare type SpecialUser = {|
  ...$Exact<User>,
  nice: number,
|};
"
`;

exports[`should handle interface inheritance 4`] = `
"declare interface User {
  firstName: string;
}
declare type SpecialUser = { nice: number, ... } & User;
"
`;

exports[`should handle interface merging 1`] = `
"declare interface User {
  firstName: string;
  lastName: string;
  username: string;
}
"
`;

exports[`should handle interface merging 2`] = `
"declare type User = {
  firstName: string,
  lastName: string,
  username: string,
  ...
};
"
`;

exports[`should handle mutli-extends pattern 1`] = `
"declare interface Shape {
  color: string;
}
declare interface PenStroke {
  penWidth: number;
}
declare type Square = { sideLength: number, ... } & Shape & PenStroke;
"
`;

exports[`should handle single interface 1`] = `
"declare interface User {
  firstName: string;
}
"
`;

exports[`should handle single interface 2`] = `
"declare type User = { firstName: string, ... };
"
`;

exports[`should handle toString property name 1`] = `
"declare interface A {
  toString(): string;
}
"
`;

exports[`should handle typed array binding pattern 1`] = `
"declare interface ArrayBinding {
  (): void;
  (x: []): void;
  (x: [string, number]): void;
}
"
`;

exports[`should handle typed object binding pattern 1`] = `
"declare interface ObjectBinding {
  (): void;
  (x: any): void;
  (x: {
    a: string,
    b: number,
    ...
  }): void;
}
"
`;

exports[`should handle untyped array binding pattern 1`] = `
"declare interface ArrayBinding {
  (): void;
  (x: any): void;
  (x: any): void;
}
"
`;

exports[`should handle untyped object binding pattern 1`] = `
"declare interface ObjectBinding {
  (): void;
  (x: { ... }): void;
  (x: {
    a: any,
    b: any,
    ...
  }): void;
}
"
`;

exports[`should remove generic defaults in call signature 1`] = `
"declare interface AbstractLevelDOWN<K, V> {}
declare interface AbstractLevelDOWNConstructor {
  <K, V>(location: string): AbstractLevelDOWN<K, V>;
}
"
`;

exports[`should remove this in call signature 1`] = `
"declare interface Arc<This, Datum> {
  (d: Datum, ...args: any[]): string | null;
}
declare interface D<This, Datum> {
  new(d: Datum, ...args: any[]): void;
}
declare interface C<This, Datum> {
  (d: Datum, ...args: any[]): any;
}
"
`;

exports[`should support call signature 1`] = `
"declare interface ObjectSchema<T> {}
declare interface ObjectSchemaDefinition<T> {}
declare interface ObjectSchemaConstructor {
  <T: { [key: string]: any }>(
    fields?: ObjectSchemaDefinition<T>
  ): ObjectSchema<T>;
  new(): ObjectSchema<{ ... }>;
}
"
`;

exports[`should support omitting generic defaults in types, classes, interfaces 1`] = `
"declare interface Foo<T = Symbol, U = number> {}
declare type FooBar = { ... } & Foo<>;
declare type Bar<T = number, U = string> = { ... };
declare class Baz<T = string, U = number> {}
declare var a: Foo<>;
declare var b: Bar<>;
declare var c: Baz<>;
declare var d: Foo<any>;
declare var e: Bar<any>;
declare var f: Baz<any>;
"
`;

exports[`should support optional methods 1`] = `
"declare interface Example<State> {
  required<R>(value: any, state: State): true;
  optional?: <R>(value: any, state: State) => false;
}
"
`;

exports[`should support readonly modifier 1`] = `
"declare interface Helper {
  +name: string;
  +callback: () => void;
}
"
`;
